### PR TITLE
add decorates to service definition

### PR DESCRIPTION
**Pull request by wodka**
implement decorators - http://symfony.com/doc/2.8/components/dependency_injection/advanced.html#decorating-services

and implement deprecated :)

fix for: https://github.com/schmittjoh/JMSDiExtraBundle/issues/230

### CODE REVIEW

File: Metadata/MetadataConverter.php [hunk 1]
```diff
@@ -64,6 +64,14 @@ public function convert(ClassHierarchyMetadata $metadata)
             $definition->setTags($classMetadata->tags);
             $definition->setProperties($classMetadata->properties);
 
+            if (null !== $classMetadata->decorates && method_exists($definition, 'setDecoratedService')) {
+                $definition->setDecoratedService($classMetadata->decorates, $classMetadata->decoration_inner_name);
```
Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: What do you think about throwing an exception if this methods are available ?

**Comment by Ener-Getick** [comment]
Can add a test to avoid regressions please?

File: Metadata/MetadataConverter.php (hunk 1 above)
Comment (wodka) [inline Metadata/MetadataConverter.php]: wouldn't do that - in a bundle I add the deprecation / decoration annotation, the bundle might work with symfony 2.3, and is now included in a project using 2.3 -> would throw the exception then

both decorating and deprecation support was added in 2.8

Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: Why would you want a weird behavior in your project ? 
That would mean that when using 2.8 your service will extend another one and in 2.3 it won't...

Comment (wodka) [inline Metadata/MetadataConverter.php]: sry mixed 2 things in my head -> exception makes sense for the decorates annotation!

only for the deprecation warning I would not add an exeption

Comment (Ener-Getick) [inline Metadata/MetadataConverter.php]: I agree this exception should only be thrown when using ``decorates``.

**Comment by wodka** [comment]
I just added some tests

File: Tests/Metadata/Driver/AnnotationDriverTest.php [hunk 2]
```diff
@@ -37,6 +37,16 @@ public function testCustomAnnotationOnClass()
...
     }
 
+    public function testServiceAnnotations()
+    {
+        $metadata = $this->getDriver()->loadMetadataForClass(new \ReflectionClass('JMS\DiExtraBundle\Tests\Metadata\Driver\Fixture\Service'));
+        $this->assertEquals('test.service', @$metadata->id);
+        $this->assertEquals('test.service', @$metadata->decorates);
+        $this->assertEquals('original.test.service', @$metadata->decoration_inner_name);
+        $this->assertEquals('use new.test.service instead', @$metadata->deprecated);
+        $this->assertEquals(false, @$metadata->public);
```
Comment (Ener-Getick) [inline Tests/Metadata/Driver/AnnotationDriverTest.php]: why do you use ``@`` ?

Comment (wodka) [inline Tests/Metadata/Driver/AnnotationDriverTest.php]: copied the wrong line ... I just removed it

**Comment by Ener-Getick** [comment]
Merged in  388dc58. Thank you @wodka

**Pull request closed without merging**